File: MuffinMealPrep/src/main/java/com/skilldistillery/meals/services/MealReviewService.java
package com.skilldistillery.meals.services;

import java.util.List;

import com.skilldistillery.meals.entities.Meal;
import com.skilldistillery.meals.entities.MealReview;
import com.skilldistillery.meals.entities.MealReviewId;
import com.skilldistillery.meals.entities.User;

public interface MealReviewService {

//	MealReview createMealReview(MealReview mealReview, Meal meal, User user);

	List<MealReview> getMealReviewsByMealId(int mealId);

	MealReview findMealReviewById(MealReviewId id);

	MealReview createMealReview(MealReview mealReview, int mealId, String username);

}

File: ngMuffinMealPrep/src/app/services/meal-review.service.ts
// src/app/services/meal-review.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable, catchError, throwError } from 'rxjs';
import { MealReview } from '../models/meal-review';
import { AuthService } from './auth.service';
import { environment } from 'src/environments/environment';


@Injectable({
  providedIn: 'root'
})
export class MealReviewService {
  private url = environment.baseUrl + 'api/meals';

  constructor(private http: HttpClient, private auth: AuthService) { }

  createMealReview(mealReview: MealReview, mealId: number, userId: number): Observable<MealReview> {
    console.log('Sending JSON payload:', JSON.stringify(mealReview));
    console.log(mealId)
    console.log(userId)
    return this.http.post<MealReview>(`${this.url}/${mealId}/mealReviews`, mealReview, this.getHttpOptions()).pipe(
      catchError((err: any) => {
        console.log(err);
        return throwError(
          () => new Error('MealReviewService.createMealReview(): error creating meal review: ' + err)
        );
      })
    );
  }

  getMealReviewsByMealId(mealId: number): Observable<MealReview[]> {
    return this.http.get<MealReview[]>(`${this.url}/${mealId}/mealReviews`, this.getHttpOptions()).pipe(
      catchError((err: any) => {
        console.log(err);
        return throwError(
          () => new Error('MealReviewService.getMealReviewsByMealId(): error fetching meal reviews by mealId: ' + err)
        );
      })
    );
  }



  getHttpOptions() {
    let options = {
      headers: {
        Authorization: 'Basic ' + this.auth.getCredentials(),
        'X-Requested-With': 'XMLHttpRequest',
      },
    };
    return options;
  }
}

